File: routes/planner.py
"""
routes/planner.py
──────────────────
Trip planning routes: itinerary generation, packing list, family planner, PDF export.
"""

import json
import logging
from flask import Blueprint, render_template, request, jsonify, session, send_file
import io

from services.watsonx_service import (
    generate_itinerary,
    get_travel_tips,
    generate_packing_list,
    plan_family_trip,
)
from services.budget_service import estimate_budget, breakdown_to_dict
from utils.helpers import sanitize_text, new_id
from utils.pdf_export import generate_itinerary_pdf

logger = logging.getLogger(__name__)
planner_bp = Blueprint("planner", __name__)


@planner_bp.route("/planner")
def planner():
    profile = session.get("profile", {})
    return render_template("planner.html", profile=profile)


@planner_bp.route("/api/itinerary/generate", methods=["POST"])
def api_generate_itinerary():
    data = request.get_json(silent=True) or {}
    destination = sanitize_text(data.get("destination", ""))
    if not destination:
        return jsonify({"error": "Destination is required"}), 400

    duration_days = int(data.get("duration_days", 5))
    travel_style = data.get("travel_style", "mid-range")
    traveler_type = data.get("traveler_type", "solo")
    interests = data.get("interests", [])
    budget_inr = float(data.get("budget_inr", 50000))
    start_date = data.get("start_date", "")
    travelers = int(data.get("travelers", 1))

    # AI-generated itinerary text
    itinerary_text = generate_itinerary(
        destination=destination,
        duration_days=duration_days,
        travel_style=travel_style,
        traveler_type=traveler_type,
        interests=interests,
        budget_inr=budget_inr,
    )

    # Budget breakdown: split the user's entered budget proportionally
    # using the style ratios, capped to exactly budget_inr.
    is_international = data.get("is_international", False)
    bb = estimate_budget(duration_days, travelers, travel_style, is_international)
    raw_total = bb.total
    if raw_total > 0 and budget_inr > 0:
        scale = budget_inr / raw_total
        from models.trip import BudgetBreakdown
        bb = BudgetBreakdown(
            accommodation=round(bb.accommodation * scale),
            transport=round(bb.transport * scale),
            food=round(bb.food * scale),
            sightseeing=round(bb.sightseeing * scale),
            shopping=round(bb.shopping * scale),
            miscellaneous=round(bb.miscellaneous * scale),
        )
    budget_breakdown = breakdown_to_dict(bb)

    trip = {
        "id": new_id(),
        "destination": destination,
        "duration_days": duration_days,
        "travel_style": travel_style,
        "traveler_type": traveler_type,
        "travelers": travelers,
        "start_date": start_date,
        "budget_inr": budget_inr,
        "itinerary_text": itinerary_text,
        "budget_breakdown": budget_breakdown,
        "interests": interests,
    }

    # Cache the last generated trip for PDF export
    session["last_trip"] = trip
    session.modified = True

    return jsonify({"itinerary": itinerary_text, "budget_breakdown": budget_breakdown, "trip_id": trip["id"]})


@planner_bp.route("/api/itinerary/pdf", methods=["POST"])
def api_export_pdf():
    from datetime import date as _date
    data = request.get_json(silent=True) or {}
    # Fall back to the server-side cached trip when the client sends empty data
    if not data.get("itinerary_text"):
        data = session.get("last_trip", data)
    if not data:
        return jsonify({"error": "No trip data available. Please generate an itinerary first."}), 400
    pdf_bytes = generate_itinerary_pdf(data)
    if pdf_bytes is None:
        return jsonify({"error": "PDF generation failed. Please try again."}), 503
    dest = (data.get("destination") or "Trip").replace(" ", "_")
    today = _date.today().isoformat()
    filename = f"Trip_Report_{dest}_{today}.pdf"
    return send_file(
        io.BytesIO(pdf_bytes),
        mimetype="application/pdf",
        as_attachment=True,
        download_name=filename,
    )


@planner_bp.route("/api/travel-tips", methods=["POST"])
def api_travel_tips():
    data = request.get_json(silent=True) or {}
    destination = sanitize_text(data.get("destination", ""))
    traveler_type = data.get("traveler_type", "solo")
    travel_month = data.get("travel_month", "January")
    if not destination:
        return jsonify({"error": "Destination is required"}), 400
    tips = get_travel_tips(destination, traveler_type, travel_month)
    return jsonify({"tips": tips})


@planner_bp.route("/api/packing-list", methods=["POST"])
def api_packing_list():
    data = request.get_json(silent=True) or {}
    destination = sanitize_text(data.get("destination", ""))
    duration_days = int(data.get("duration_days", 5))
    activities = data.get("activities", [])
    travel_month = data.get("travel_month", "January")
    if not destination:
        return jsonify({"error": "Destination required"}), 400
    checklist = generate_packing_list(destination, duration_days, activities, travel_month)
    return jsonify({"checklist": checklist})


@planner_bp.route("/api/family-trip", methods=["POST"])
def api_family_trip():
    data = request.get_json(silent=True) or {}
    destination = sanitize_text(data.get("destination", ""))
    duration_days = int(data.get("duration_days", 5))
    adults = int(data.get("adults", 2))
    children_ages = data.get("children_ages", [])
    budget_inr = float(data.get("budget_inr", 80000))
    if not destination:
        return jsonify({"error": "Destination required"}), 400
    result = plan_family_trip(destination, duration_days, adults, children_ages, budget_inr)
    return jsonify({"plan": result})

File: static/js/planner.js
/**
 * planner.js — Trip Planner page logic
 */

// ── Interest tags ────────────────────────────────────────────────────────────
document.querySelectorAll('#interestTags .interest-tag').forEach(tag => {
  tag.addEventListener('click', () => tag.classList.toggle('selected'));
});

// ── Travel Style radio visual feedback ──────────────────────────────────────
document.querySelectorAll('input[name="travelStyle"]').forEach(radio => {
  radio.addEventListener('change', () => {
    document.querySelectorAll('.style-option').forEach(opt => opt.classList.remove('checked'));
    radio.closest('.style-option').classList.add('checked');
  });
});

// ── Show family panel when traveler type is "family" ─────────────────────────
document.getElementById('travelerType')?.addEventListener('change', function() {
  const panel = document.getElementById('familyPanel');
  panel.style.display = this.value === 'family' ? 'block' : 'none';
});

// ── Pre-fill destination from URL param ──────────────────────────────────────
const urlParams = new URLSearchParams(window.location.search);
const destParam = urlParams.get('dest');
if (destParam) {
  const destInput = document.getElementById('destination');
  if (destInput) destInput.value = destParam;
}

// ── Date sync ────────────────────────────────────────────────────────────────
document.getElementById('startDate')?.addEventListener('change', function() {
  const days = parseInt(document.getElementById('durationDays').value) || 5;
  const start = new Date(this.value);
  if (!isNaN(start)) {
    start.setDate(start.getDate() + days);
    document.getElementById('endDate').value = start.toISOString().split('T')[0];
  }
});

document.getElementById('durationDays')?.addEventListener('change', function() {
  const startVal = document.getElementById('startDate').value;
  if (!startVal) return;
  const start = new Date(startVal);
  start.setDate(start.getDate() + parseInt(this.value));
  document.getElementById('endDate').value = start.toISOString().split('T')[0];
});

// ── Budget chart reference ───────────────────────────────────────────────────
let budgetChartInstance = null;

// ── Generate Itinerary ───────────────────────────────────────────────────────
document.getElementById('generateBtn')?.addEventListener('click', generateItinerary);

async function generateItinerary() {
  const destination = document.getElementById('destination').value.trim();
  if (!destination) {
    showToast('Please enter a destination.', 'warning');
    return;
  }

  const selectedInterests = [...document.querySelectorAll('#interestTags .interest-tag.selected')]
    .map(t => t.dataset.tag);

  const payload = {
    destination,
    duration_days: parseInt(document.getElementById('durationDays').value) || 5,
    travelers: parseInt(document.getElementById('travelers').value) || 1,
    start_date: document.getElementById('startDate').value,
    end_date: document.getElementById('endDate').value,
    travel_style: document.querySelector('input[name="travelStyle"]:checked')?.value || 'mid-range',
    traveler_type: document.getElementById('travelerType').value,
    budget_inr: parseFloat(document.getElementById('budgetInr').value) || 50000,
    is_international: document.getElementById('isInternational').checked,
    interests: selectedInterests,
  };

  setLoadingState(true);

  try {
    const res = await fetch('/api/itinerary/generate', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
    const data = await res.json();

    if (data.error) { showToast(data.error, 'danger'); return; }

    // Render itinerary
    document.getElementById('itineraryText').innerHTML = formatAIText(data.itinerary);
    document.getElementById('plannerPlaceholder').style.display = 'none';
    document.getElementById('itineraryResult').style.display = 'block';

    // Render budget breakdown
    renderBudgetBreakdown(data.budget_breakdown, payload);

    // Store trip data for saving
    window._currentTrip = { ...payload, itinerary_text: data.itinerary, budget_breakdown: data.budget_breakdown };

    showToast('Itinerary generated successfully!', 'success');
  } catch (err) {
    showToast('Failed to generate itinerary. Check AI service connection.', 'danger');
  } finally {
    setLoadingState(false);
  }
}

// ── Render Budget Breakdown ───────────────────────────────────────────────────
function renderBudgetBreakdown(bb, payload) {
  if (!bb) return;

  const categories = ['accommodation','transport','food','sightseeing','shopping','miscellaneous'];
  const labels = ['Accommodation','Transport','Food','Sightseeing','Shopping','Misc'];
  const colors = ['#1a56db','#7c3aed','#059669','#d97706','#ef4444','#0891b2'];

  // Read CSS variables at render time so colors adapt to the active theme
  const cs = getComputedStyle(document.documentElement);
  const themeColors = [
    cs.getPropertyValue('--primary').trim()  || colors[0],
    '#7c3aed',
    cs.getPropertyValue('--success').trim()  || colors[2],
    cs.getPropertyValue('--warning').trim()  || colors[3],
    cs.getPropertyValue('--danger').trim()   || colors[4],
    cs.getPropertyValue('--info').trim()     || colors[5],
  ];

  // Stat mini-cards
  const cardsContainer = document.getElementById('budgetCards');
  cardsContainer.innerHTML = '';
  categories.forEach((cat, i) => {
    const card = document.createElement('div');
    card.className = 'col-6 col-sm-4';
    card.innerHTML = `
      <div class="budget-stat-card">
        <div class="budget-stat-label">${labels[i]}</div>
        <div class="budget-stat-value" style="color:${themeColors[i]}">₹${formatNum(bb[cat])}</div>
      </div>`;
    cardsContainer.appendChild(card);
  });

  // Total display
  const totalCard = document.createElement('div');
  totalCard.className = 'col-12 mt-1';
  totalCard.innerHTML = `
    <div class="total-budget-display">
      Total Estimated Cost: ₹${formatNum(bb.total)} for ${payload.travelers} traveler(s)
    </div>`;
  cardsContainer.appendChild(totalCard);

  // Doughnut chart
  if (budgetChartInstance) budgetChartInstance.destroy();
  const ctx = document.getElementById('budgetChart').getContext('2d');
  budgetChartInstance = new Chart(ctx, {
    type: 'doughnut',
    data: {
      labels: labels,
      datasets: [{
        data: categories.map(c => bb[c]),
        backgroundColor: colors,
        borderWidth: 2,
        borderColor: getComputedStyle(document.documentElement).getPropertyValue('--surface').trim() || '#fff',
      }],
    },
    options: {
      responsive: true,
      cutout: '60%',
      plugins: {
        legend: {
          position: 'bottom',
          labels: {
            color: getComputedStyle(document.documentElement).getPropertyValue('--text').trim(),
            font: { size: 12 },
            padding: 14,
          },
        },
        tooltip: {
          callbacks: {
            label: ctx => ` ₹${formatNum(ctx.parsed)} (${((ctx.parsed / bb.total) * 100).toFixed(1)}%)`,
          },
        },
      },
    },
  });
}

// ── Get Travel Tips ───────────────────────────────────────────────────────────
document.getElementById('tipsBtn')?.addEventListener('click', async () => {
  const destination = document.getElementById('destination').value.trim();
  if (!destination) { showToast('Enter a destination first.', 'warning'); return; }

  const tipsResult = document.getElementById('tipsResult');
  tipsResult.style.display = 'block';
  document.getElementById('tipsText').innerHTML = '<div class="text-center py-3"><div class="spinner-border text-primary"></div></div>';

  try {
    const res = await fetch('/api/travel-tips', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        destination,
        traveler_type: document.getElementById('travelerType').value,
        travel_month: new Date().toLocaleString('default', { month: 'long' }),
      }),
    });
    const data = await res.json();
    document.getElementById('tipsText').innerHTML = formatAIText(data.tips || data.error);
  } catch {
    document.getElementById('tipsText').innerHTML = '<p class="text-danger">Failed to load tips.</p>';
  }
});

// ── Get Packing List ──────────────────────────────────────────────────────────
document.getElementById('packingBtn')?.addEventListener('click', async () => {
  const destination = document.getElementById('destination').value.trim();
  if (!destination) { showToast('Enter a destination first.', 'warning'); return; }

  const packingResult = document.getElementById('packingResult');
  packingResult.style.display = 'block';
  document.getElementById('packingText').innerHTML = '<div class="text-center py-3"><div class="spinner-border text-info"></div></div>';

  try {
    const res = await fetch('/api/packing-list', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        destination,
        duration_days: parseInt(document.getElementById('durationDays').value) || 5,
        activities: [...document.querySelectorAll('#interestTags .interest-tag.selected')].map(t => t.dataset.tag),
        travel_month: new Date().toLocaleString('default', { month: 'long' }),
      }),
    });
    const data = await res.json();
    document.getElementById('packingText').innerHTML = formatAIText(data.checklist || data.error);
  } catch {
    document.getElementById('packingText').innerHTML = '<p class="text-danger">Failed to load packing list.</p>';
  }
});

// ── Family Trip ────────────────────────────────────────────────────────────────
document.getElementById('familyPlanBtn')?.addEventListener('click', async () => {
  const destination = document.getElementById('destination').value.trim();
  if (!destination) { showToast('Enter a destination first.', 'warning'); return; }

  const agesRaw = document.getElementById('childrenAges').value;
  const childrenAges = agesRaw.split(',').map(a => parseInt(a.trim())).filter(a => !isNaN(a));

  setLoadingState(true);
  try {
    const res = await fetch('/api/family-trip', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        destination,
        duration_days: parseInt(document.getElementById('durationDays').value),
        adults: parseInt(document.getElementById('adultsCount').value),
        children_ages: childrenAges,
        budget_inr: parseFloat(document.getElementById('budgetInr').value),
      }),
    });
    const data = await res.json();
    document.getElementById('plannerPlaceholder').style.display = 'none';
    document.getElementById('itineraryResult').style.display = 'block';
    document.getElementById('itineraryText').innerHTML = formatAIText(data.plan || data.error);
    showToast('Family trip plan ready!', 'success');
  } catch {
    showToast('Failed to generate family plan.', 'danger');
  } finally {
    setLoadingState(false);
  }
});

// ── Export PDF ────────────────────────────────────────────────────────────────
document.getElementById('exportPdfBtn')?.addEventListener('click', async () => {
  if (!window._currentTrip) {
    showToast('Please generate an itinerary before exporting PDF.', 'warning');
    return;
  }
  showToast('Generating PDF…', 'info', 3000);
  try {
    const res = await fetch('/api/itinerary/pdf', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(window._currentTrip),
    });
    if (!res.ok) {
      let errMsg = 'PDF export failed. Please try again.';
      try { const err = await res.json(); errMsg = err.error || errMsg; } catch {}
      showToast(errMsg, 'danger', 5000);
      return;
    }
    const blob = await res.blob();
    if (!blob || blob.size === 0) {
      showToast('PDF export failed: empty response. Please try again.', 'danger', 5000);
      return;
    }
    // Build filename: Trip_Report_<Destination>_<YYYY-MM-DD>.pdf
    const dest = (window._currentTrip.destination || 'Trip').replace(/\s+/g, '_');
    const today = new Date().toISOString().split('T')[0];
    const filename = `Trip_Report_${dest}_${today}.pdf`;
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
    showToast('PDF downloaded successfully!', 'success');
  } catch (err) {
    console.error('PDF export error:', err);
    showToast('PDF export failed. Check your connection and try again.', 'danger', 5000);
  }
});

// ── Save Trip ─────────────────────────────────────────────────────────────────
document.getElementById('saveTripBtn')?.addEventListener('click', async () => {
  if (!window._currentTrip) {
    showToast('Generate an itinerary first.', 'warning');
    return;
  }
  try {
    const res = await fetch('/api/trip/save', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(window._currentTrip),
    });
    const data = await res.json();
    if (data.status === 'ok') showToast('Trip saved successfully!', 'success');
  } catch {
    showToast('Failed to save trip.', 'danger');
  }
});

// ── Helpers ───────────────────────────────────────────────────────────────────
function setLoadingState(loading) {
  document.getElementById('loadingSkeleton').style.display = loading ? 'block' : 'none';
  if (loading) {
    document.getElementById('itineraryResult').style.display = 'none';
    document.getElementById('plannerPlaceholder').style.display = 'none';
  }
  document.getElementById('generateBtn').disabled = loading;
  document.getElementById('generateBtn').innerHTML = loading
    ? '<span class="spinner-border spinner-border-sm me-2"></span>Generating…'
    : '<i class="bi bi-magic me-2"></i>Generate Itinerary';
}

function formatNum(n) {
  return Number(n).toLocaleString('en-IN', { maximumFractionDigits: 0 });
}
